### PR TITLE
atomics: added support for GCC __atomic builtins

### PR DESCRIPTION
**Pull request by snikulov**
Added support for __atomic builtins, since __sync is deprecated https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html
Updated tests, to handle those changes.

### CODE REVIEW

File: tests/ipc_shutdown.c [hunk 1]
```diff
@@ -110,9 +111,9 @@ int main ()
     for (j = 0; j != TEST_LOOPS; ++j) {
         for (i = 0; i != TEST2_THREAD_COUNT; ++i)
             nn_thread_init (&threads [i], routine2, NULL);
-        active = TEST2_THREAD_COUNT;
+        nn_atomic_init(&active,TEST2_THREAD_COUNT);
 
-        while (active) {
+        while (nn_atomic_fetch(&active)) {
```
[the remark this answers is not in the document]
Comment (snikulov) [reply tests/ipc_shutdown.c]: Volatile was never adequate for multithreading.
So atomic_fetch is better, IMHO. Or Atomic_ flag for structure, but for C11

**Comment by snikulov** [comment]
@gdamore Yeah. Working on windows stuff.

File: src/utils/atomic.c [hunk 2]
```diff
@@ -38,12 +43,23 @@ void nn_atomic_term (struct nn_atomic *self)
...
+}
+
 uint32_t nn_atomic_inc (struct nn_atomic *self, uint32_t n)
 {
 #if defined NN_ATOMIC_WINAPI
     return (uint32_t) InterlockedExchangeAdd ((LONG*) &self->n, n);
 #elif defined NN_ATOMIC_SOLARIS
     return atomic_add_32_nv (&self->n, n) - n;
+#elif defined NN_ATOMIC_GCC_MM_BUILTINS
+    return (uint32_t) __atomic_add_fetch (&self->n, n, __ATOMIC_SEQ_CST);
```
[the remark this answers is not in the document]
Comment (snikulov) [reply src/utils/atomic.c]: @gdamore Unfortunately I've not found such nice addition.

**Comment by snikulov** [comment]
@gdamore 

> Linux code is breaking too, so there is something wrong here. Its not obvious to me what the problem is there -- it would seem to be the case that this should work from reading the code, but it was working before your change, and the issue appears to be in the chunk stuff that does use atomics.

Do you mind if it was not working before? :)
What if I add the option to run tests under threadsanitizer? 
Will it be useful?

**Comment by snikulov** [comment]
@gdamore all green now.

**Comment by gdamore** [comment]
Ok I will review...

File: tests/ipc_shutdown.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (gdamore) [reply tests/ipc_shutdown.c]: The point is that a *read* against this is *cannot ever be zero* if other threads are running.  I.e. this can be lockless/atomic-free.  The volatile bit is necessary to just make sure we check memory again and don't rely on cache.

The *decrement* on the other hand, is prone to races, and does need the volatile, as you rightly notice.

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -339,9 +339,10 @@ if (NN_STATIC_LIB)
     add_library (${PROJECT_NAME} STATIC ${NN_SOURCES})
 else ()
     add_library (${PROJECT_NAME} SHARED ${NN_SOURCES})
-    add_definitions (-DNN_SHARED_LIB)
     set_target_properties (${PROJECT_NAME} PROPERTIES
         SOVERSION "${NN_ABI_VERSION}")
+    set_target_properties (${PROJECT_NAME} PROPERTIES
+        COMPILE_DEFINITIONS "NN_SHARED_LIB")
```
Comment (gdamore) [inline src/CMakeLists.txt]: So this CMake thing is a bit different -- I'm concerned that if this doesn't work, we might wind up with a static library.   I *think* what you're trying to achieve here is correct, but I want to validate this.  Unfortunately, the failure mode for this if you got it wrong is that tests will just silently succeed.

File: tests/ipc_stress.c [hunk 4]
```diff
@@ -110,9 +109,12 @@ int main()
 
     for (i = 0; i != THREAD_COUNT; ++i)
         nn_thread_init(&cli_threads[i], client, (void *)(intptr_t)i);
+
     for (i = 0; i != THREAD_COUNT; ++i)
         nn_thread_term(&cli_threads[i]);
 
+    nn_thread_term(&srv_thread);
```
Comment (gdamore) [inline tests/ipc_stress.c]: This change looks unrelated.

**Comment by gdamore** [comment]
As far as adding thread sanitizer to this project -- no, I don't think so at this point.  I'm trying to limit the new work being done here.

I still don't really understand the point of all this trouble ... the existing code worked (admittedly the tests were racy, but at this point fixing those doesn't improve the quality of the code for the project at large.)

Its very likely that there won't be a follow up "release" of nanomsg for this code to be integrated into, as well (because I'm focused on the next generation project) -- i.e. this is in sustaining mode only, and absent another more serious problem, there probably won't be another release.

File: src/CMakeLists.txt (hunk 3 above)
Comment (snikulov) [reply src/CMakeLists.txt]: @gdamore anyway this is correct.
NN_SHARED_LIB required only for building the library. 
So I reduced the scope of compile definition for the specific target.

File: tests/ipc_stress.c (hunk 4 above)
Comment (snikulov) [reply tests/ipc_stress.c]: @gdamore suggesting another pull request? 
For me, it is related, because I've checked changes with thread sanitizer. 
We have thread leak in this case.

File: src/utils/atomic.c [hunk 5]
```diff
@@ -25,7 +25,12 @@
 
 void nn_atomic_init (struct nn_atomic *self, uint32_t n)
 {
+#if defined NN_ATOMIC_GCC_MM_BUILTINS
+    __atomic_store_n (&self->n, n, __ATOMIC_SEQ_CST);
+#else
     self->n = n;
+#endif
```
[the remark this answers is not in the document]
Comment (snikulov) [reply src/utils/atomic.c]: @gdamore perhaps, but nothing to stop use it multithreaded environment. And in my opinion, it is better to have some guard for that.

[the remark this answers is not in the document]
Comment (gdamore) [reply src/utils/atomic.c]: No, by definition the *semantic* is only appropriate to single threaded use.  Multithreaded initialization of the same value is simply nonsensical.

File: tests/ipc_stress.c (hunk 4 above)
Comment (gdamore) [reply tests/ipc_stress.c]: Right.  That's another issue.  It could be a different commit, I suppose, which would be slightly better... I'm willing to be a bit lax on that though.

**Comment by gdamore** [comment]
The larger question I have here -- which why are we bothering with this -- still stands.

**Comment by snikulov** [comment]
@gdamore because I evaluating nanomsg to use in production :) 
I'll not wait for nng, sorry.
Following fixes were related quality of lib and test suite.
If you as maintainer doesn't care, I'll not either.

**Comment by skyformat99** [comment]
This PR has been modified too much, not good. 
And it has to do with the version of GCC, 
which is bad for embedded development and supports the author's point of view.

**Comment by gdamore** [comment]
I'm closing this -- there is no compelling need to adopt the "new" GCC'isms for atomics, since the old ones aren't going away, and there are other concerns with this change set.
